File: Sources/UsageBarLocalAPI/StrictHTTPRequestParser.swift
import Foundation

enum LocalUsageAPIHTTPStatus: Int, Equatable {
    case ok = 200
    case badRequest = 400
    case notFound = 404
    case methodNotAllowed = 405
    case contentTooLarge = 413
    case uriTooLong = 414
    case tooManyRequests = 429
    case requestHeaderFieldsTooLarge = 431
    case serviceUnavailable = 503
    case httpVersionNotSupported = 505

    var reasonPhrase: String {
        switch self {
        case .ok: "OK"
        case .badRequest: "Bad Request"
        case .notFound: "Not Found"
        case .methodNotAllowed: "Method Not Allowed"
        case .contentTooLarge: "Content Too Large"
        case .uriTooLong: "URI Too Long"
        case .tooManyRequests: "Too Many Requests"
        case .requestHeaderFieldsTooLarge: "Request Header Fields Too Large"
        case .serviceUnavailable: "Service Unavailable"
        case .httpVersionNotSupported: "HTTP Version Not Supported"
        }
    }
}

enum StrictHTTPRequestParseResult: Equatable {
    case incomplete
    case accepted
    case rejected(LocalUsageAPIHTTPStatus)
}

struct StrictHTTPRequestParser {
    static let maximumRequestLineBytes = 1_024
    static let maximumHeaderSectionBytes = 8_192
    static let maximumHeaderCount = 32
    static let maximumHeaderLineBytes = 1_024
    static let maximumBufferedRequestBytes = maximumRequestLineBytes + maximumHeaderSectionBytes + 1

    let expectedHost: String

    func parse(_ data: Data) -> StrictHTTPRequestParseResult {
        if containsBareLineFeed(in: data) || containsMalformedCarriageReturn(in: data) {
            return .rejected(.badRequest)
        }

        guard let requestLineEnd = data.firstRange(of: Self.crlf)?.lowerBound else {
            return data.count >= Self.maximumRequestLineBytes
                ? .rejected(.uriTooLong)
                : .incomplete
        }

        let requestLineBytes = requestLineEnd + Self.crlf.count
        guard requestLineBytes <= Self.maximumRequestLineBytes else {
            return .rejected(.uriTooLong)
        }

        let headerStart = requestLineBytes
        guard let headerTerminator = data.range(of: Self.headerTerminator, in: requestLineEnd..<data.count) else {
            return data.count - headerStart >= Self.maximumHeaderSectionBytes
                ? .rejected(.requestHeaderFieldsTooLarge)
                : .incomplete
        }

        let headerEnd = headerTerminator.upperBound
        guard headerEnd - headerStart <= Self.maximumHeaderSectionBytes else {
            return .rejected(.requestHeaderFieldsTooLarge)
        }

        guard let requestLine = asciiString(data[0..<requestLineEnd]) else {
            return .rejected(.badRequest)
        }
        let requestParts = requestLine.split(separator: " ", omittingEmptySubsequences: false)
        guard requestParts.count == 3,
              !requestParts[0].isEmpty,
              !requestParts[1].isEmpty,
              !requestParts[2].isEmpty,
              requestParts.allSatisfy({ !$0.contains("\t") }),
              isToken(String(requestParts[0]))
        else {
            return .rejected(.badRequest)
        }

        let method = String(requestParts[0])
        let target = String(requestParts[1])
        let version = String(requestParts[2])
        guard isSyntacticallyValidHTTPVersion(version) else {
            return .rejected(.badRequest)
        }
        guard version == "HTTP/1.1" else {
            return .rejected(.httpVersionNotSupported)
        }

        let headerPayloadEnd = headerTerminator.lowerBound
        var cursor = headerStart
        var headers: [(name: String, value: String)] = []
        while cursor < headerPayloadEnd {
            guard let lineEnd = data.range(of: Self.crlf, in: cursor..<headerTerminator.upperBound)?.lowerBound,
                  lineEnd <= headerPayloadEnd
            else {
                return .rejected(.badRequest)
            }
            let lineByteCount = lineEnd - cursor + Self.crlf.count
            guard lineByteCount <= Self.maximumHeaderLineBytes else {
                return .rejected(.requestHeaderFieldsTooLarge)
            }
            guard let line = asciiString(data[cursor..<lineEnd]),
                  !line.isEmpty,
                  !line.first!.isWhitespace,
                  let colon = line.firstIndex(of: ":")
            else {
                return .rejected(.badRequest)
            }

            let name = String(line[..<colon])
            guard isToken(name) else {
                return .rejected(.badRequest)
            }
            var value = String(line[line.index(after: colon)...])
            if value.first == " " {
                value.removeFirst()
            }
            guard !value.isEmpty || name.caseInsensitiveCompare("Content-Length") == .orderedSame,
                  !value.hasPrefix(" "),
                  !value.hasSuffix(" "),
                  value.utf8.allSatisfy({ $0 >= 0x20 && $0 <= 0x7e })
            else {
                return .rejected(.badRequest)
            }
            headers.append((name, value))
            guard headers.count <= Self.maximumHeaderCount else {
                return .rejected(.requestHeaderFieldsTooLarge)
            }
            cursor = lineEnd + Self.crlf.count
        }

        let hosts = values(named: "Host", in: headers)
        guard hosts.count == 1, hosts[0] == expectedHost else {
            return .rejected(.badRequest)
        }
        guard values(named: "Origin", in: headers).isEmpty else {
            return .rejected(.badRequest)
        }
        let fetchSites = values(named: "Sec-Fetch-Site", in: headers)
        guard fetchSites.count <= 1, fetchSites.first == nil || fetchSites.first == "none" else {
            return .rejected(.badRequest)
        }

        guard values(named: "Transfer-Encoding", in: headers).isEmpty else {
            return .rejected(.badRequest)
        }
        let contentLengths = values(named: "Content-Length", in: headers)
        guard contentLengths.count <= 1 else {
            return .rejected(.badRequest)
        }
        if let contentLength = contentLengths.first {
            guard !contentLength.isEmpty,
                  contentLength.allSatisfy({ $0.isASCII && $0.isNumber })
            else {
                return .rejected(.badRequest)
            }
            if contentLength == "0" {
                // The only body framing accepted by the frozen protocol.
            } else if contentLength.first == "0" {
                return .rejected(.badRequest)
            } else if UInt64(contentLength) != nil {
                return .rejected(.contentTooLarge)
            } else {
                return .rejected(.badRequest)
            }
        }

        guard data.count == headerEnd else {
            return .rejected(.badRequest)
        }
        guard method == "GET" else {
            return .rejected(.methodNotAllowed)
        }
        guard target == "/v1/usage" else {
            return .rejected(.notFound)
        }
        return .accepted
    }

    private func values(named expectedName: String, in headers: [(name: String, value: String)]) -> [String] {
        headers.compactMap { header in
            header.name.caseInsensitiveCompare(expectedName) == .orderedSame ? header.value : nil
        }
    }

    private func isToken(_ value: String) -> Bool {
        !value.isEmpty && value.utf8.allSatisfy { byte in
            switch byte {
            case 0x21, 0x23...0x27, 0x2a, 0x2b, 0x2d, 0x2e,
                 0x30...0x39, 0x41...0x5a, 0x5e...0x7a, 0x7c, 0x7e:
                true
            default:
                false
            }
        }
    }

    private func isSyntacticallyValidHTTPVersion(_ value: String) -> Bool {
        let parts = value.split(separator: "/", omittingEmptySubsequences: false)
        guard parts.count == 2, parts[0] == "HTTP" else { return false }
        let numbers = parts[1].split(separator: ".", omittingEmptySubsequences: false)
        return numbers.count == 2
            && !numbers[0].isEmpty
            && !numbers[1].isEmpty
            && numbers[0].allSatisfy(\.isNumber)
            && numbers[1].allSatisfy(\.isNumber)
    }

    private func containsBareLineFeed(in data: Data) -> Bool {
        for index in data.indices where data[index] == 0x0a {
            if index == data.startIndex || data[data.index(before: index)] != 0x0d {
                return true
            }
        }
        return false
    }

    private func containsMalformedCarriageReturn(in data: Data) -> Bool {
        for index in data.indices where data[index] == 0x0d {
            let next = data.index(after: index)
            if next < data.endIndex && data[next] != 0x0a {
                return true
            }
        }
        return false
    }

    private func asciiString(_ bytes: Data.SubSequence) -> String? {
        String(data: Data(bytes), encoding: .ascii)
    }

    private static let crlf = Data([0x0d, 0x0a])
    private static let headerTerminator = Data([0x0d, 0x0a, 0x0d, 0x0a])
}

private extension Data {
    func firstRange(of needle: Data) -> Range<Int>? {
        range(of: needle, options: [], in: startIndex..<endIndex)
    }
}

File: windows/src/UsageBar.Windows.Infrastructure/LocalApi/StrictHttpRequestParser.cs
using System.Globalization;
using System.Text;

namespace UsageBar.Windows.Infrastructure.LocalApi;

internal enum StrictHttpRequestParseResultKind
{
    Incomplete,
    Accepted,
    Rejected
}

internal readonly record struct StrictHttpRequestParseResult(
    StrictHttpRequestParseResultKind Kind,
    LocalUsageApiHttpStatus? Status = null)
{
    public static StrictHttpRequestParseResult Incomplete() =>
        new(StrictHttpRequestParseResultKind.Incomplete);

    public static StrictHttpRequestParseResult Accepted() =>
        new(StrictHttpRequestParseResultKind.Accepted);

    public static StrictHttpRequestParseResult Rejected(LocalUsageApiHttpStatus status) =>
        new(StrictHttpRequestParseResultKind.Rejected, status);
}

internal sealed class StrictHttpRequestParser
{
    public const int MaximumRequestLineBytes = 1024;
    public const int MaximumHeaderSectionBytes = 8192;
    public const int MaximumHeaderCount = 32;
    public const int MaximumHeaderLineBytes = 1024;
    public const int MaximumBufferedRequestBytes =
        MaximumRequestLineBytes + MaximumHeaderSectionBytes + 1;

    private readonly string _expectedHost;

    public StrictHttpRequestParser(string expectedHost)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(expectedHost);
        _expectedHost = expectedHost;
    }

    public StrictHttpRequestParseResult Parse(ReadOnlySpan<byte> data)
    {
        if (ContainsBareLineFeed(data) || ContainsMalformedCarriageReturn(data))
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
        }

        var requestLineEnd = IndexOf(data, CrLf, 0);
        if (requestLineEnd < 0)
        {
            return data.Length >= MaximumRequestLineBytes
                ? StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.UriTooLong)
                : StrictHttpRequestParseResult.Incomplete();
        }

        var requestLineBytes = requestLineEnd + CrLf.Length;
        if (requestLineBytes > MaximumRequestLineBytes)
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.UriTooLong);
        }

        var headerStart = requestLineBytes;
        var headerTerminator = IndexOf(data, HeaderTerminator, requestLineEnd);
        if (headerTerminator < 0)
        {
            return data.Length - headerStart >= MaximumHeaderSectionBytes
                ? StrictHttpRequestParseResult.Rejected(
                    LocalUsageApiHttpStatus.RequestHeaderFieldsTooLarge)
                : StrictHttpRequestParseResult.Incomplete();
        }

        var headerEnd = headerTerminator + HeaderTerminator.Length;
        if (headerEnd - headerStart > MaximumHeaderSectionBytes)
        {
            return StrictHttpRequestParseResult.Rejected(
                LocalUsageApiHttpStatus.RequestHeaderFieldsTooLarge);
        }

        if (!TryAscii(data[..requestLineEnd], out var requestLine))
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
        }

        var requestParts = requestLine.Split(' ', StringSplitOptions.None);
        if (requestParts.Length != 3 ||
            requestParts.Any(string.IsNullOrEmpty) ||
            requestParts.Any(part => part.Contains('\t')) ||
            !IsToken(requestParts[0]))
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
        }

        var method = requestParts[0];
        var target = requestParts[1];
        var version = requestParts[2];
        if (!IsSyntacticallyValidHttpVersion(version))
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
        }

        if (!string.Equals(version, "HTTP/1.1", StringComparison.Ordinal))
        {
            return StrictHttpRequestParseResult.Rejected(
                LocalUsageApiHttpStatus.HttpVersionNotSupported);
        }

        var headers = new List<(string Name, string Value)>();
        var cursor = headerStart;
        while (cursor < headerTerminator)
        {
            var lineEnd = IndexOf(data, CrLf, cursor);
            if (lineEnd < cursor || lineEnd > headerTerminator)
            {
                return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
            }

            var lineByteCount = lineEnd - cursor + CrLf.Length;
            if (lineByteCount > MaximumHeaderLineBytes)
            {
                return StrictHttpRequestParseResult.Rejected(
                    LocalUsageApiHttpStatus.RequestHeaderFieldsTooLarge);
            }

            if (!TryAscii(data[cursor..lineEnd], out var line) ||
                line.Length == 0 ||
                char.IsWhiteSpace(line[0]))
            {
                return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
            }

            var colon = line.IndexOf(':');
            if (colon < 0)
            {
                return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
            }

            var name = line[..colon];
            if (!IsToken(name))
            {
                return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
            }

            var value = line[(colon + 1)..];
            if (value.StartsWith(' '))
            {
                value = value[1..];
            }

            if ((value.Length == 0 && !name.Equals("Content-Length", StringComparison.OrdinalIgnoreCase)) ||
                value.StartsWith(' ') ||
                value.EndsWith(' ') ||
                value.Any(character => character is < (char)0x20 or > (char)0x7e))
            {
                return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
            }

            headers.Add((name, value));
            if (headers.Count > MaximumHeaderCount)
            {
                return StrictHttpRequestParseResult.Rejected(
                    LocalUsageApiHttpStatus.RequestHeaderFieldsTooLarge);
            }

            cursor = lineEnd + CrLf.Length;
        }

        var hosts = Values("Host", headers);
        if (hosts.Count != 1 || !string.Equals(hosts[0], _expectedHost, StringComparison.Ordinal))
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
        }

        if (Values("Origin", headers).Count != 0)
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
        }

        var fetchSites = Values("Sec-Fetch-Site", headers);
        if (fetchSites.Count > 1 ||
            (fetchSites.Count == 1 && !string.Equals(fetchSites[0], "none", StringComparison.Ordinal)))
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
        }

        if (Values("Transfer-Encoding", headers).Count != 0)
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
        }

        var contentLengths = Values("Content-Length", headers);
        if (contentLengths.Count > 1)
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
        }

        if (contentLengths.Count == 1)
        {
            var contentLength = contentLengths[0];
            if (contentLength.Length == 0 || contentLength.Any(character => character is < '0' or > '9'))
            {
                return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
            }

            if (contentLength == "0")
            {
                // The only accepted body framing.
            }
            else if (contentLength[0] == '0')
            {
                return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
            }
            else if (ulong.TryParse(contentLength, NumberStyles.None, CultureInfo.InvariantCulture, out _))
            {
                return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.ContentTooLarge);
            }
            else
            {
                return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
            }
        }

        if (data.Length != headerEnd)
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.BadRequest);
        }

        if (!string.Equals(method, "GET", StringComparison.Ordinal))
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.MethodNotAllowed);
        }

        if (!string.Equals(target, "/v1/usage", StringComparison.Ordinal))
        {
            return StrictHttpRequestParseResult.Rejected(LocalUsageApiHttpStatus.NotFound);
        }

        return StrictHttpRequestParseResult.Accepted();
    }

    private static List<string> Values(
        string expectedName,
        IEnumerable<(string Name, string Value)> headers) =>
        headers
            .Where(header => header.Name.Equals(expectedName, StringComparison.OrdinalIgnoreCase))
            .Select(header => header.Value)
            .ToList();

    private static bool IsToken(string value) =>
        value.Length > 0 && value.All(character => character switch
        {
            '!' or '#' or '$' or '%' or '&' or '\'' or '*' or '+' or '-' or '.' or
            '^' or '_' or '`' or '|' or '~' => true,
            >= '0' and <= '9' => true,
            >= 'A' and <= 'Z' => true,
            >= 'a' and <= 'z' => true,
            _ => false
        });

    private static bool IsSyntacticallyValidHttpVersion(string value)
    {
        var slash = value.Split('/', StringSplitOptions.None);
        if (slash.Length != 2 || slash[0] != "HTTP")
        {
            return false;
        }

        var numbers = slash[1].Split('.', StringSplitOptions.None);
        return numbers.Length == 2 &&
               numbers.All(number =>
                   number.Length > 0 && number.All(character => character is >= '0' and <= '9'));
    }

    private static bool ContainsBareLineFeed(ReadOnlySpan<byte> data)
    {
        for (var index = 0; index < data.Length; index += 1)
        {
            if (data[index] == 0x0a && (index == 0 || data[index - 1] != 0x0d))
            {
                return true;
            }
        }

        return false;
    }

    private static bool ContainsMalformedCarriageReturn(ReadOnlySpan<byte> data)
    {
        for (var index = 0; index < data.Length; index += 1)
        {
            if (data[index] == 0x0d && index + 1 < data.Length && data[index + 1] != 0x0a)
            {
                return true;
            }
        }

        return false;
    }

    private static bool TryAscii(ReadOnlySpan<byte> bytes, out string value)
    {
        foreach (var item in bytes)
        {
            if (item > 0x7f)
            {
                value = string.Empty;
                return false;
            }
        }

        value = Encoding.ASCII.GetString(bytes);
        return true;
    }

    private static int IndexOf(ReadOnlySpan<byte> data, ReadOnlySpan<byte> needle, int start)
    {
        if (start < 0 || start > data.Length)
        {
            return -1;
        }

        var relative = data[start..].IndexOf(needle);
        return relative < 0 ? -1 : start + relative;
    }

    private static ReadOnlySpan<byte> CrLf => new byte[] { 0x0d, 0x0a };

    private static ReadOnlySpan<byte> HeaderTerminator =>
        new byte[] { 0x0d, 0x0a, 0x0d, 0x0a };
}
